errors/errors.go

```diff
@@ -15,209 +15,310 @@ package errors
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
-	"net/http"
-
+	"github.com/dapr/kit/logger"
 	"google.golang.org/genproto/googleapis/rpc/errdetails"
-	"google.golang.org/grpc/codes"
+	grpcCodes "google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
-	"google.golang.org/protobuf/encoding/protojson"
-
-	"github.com/dapr/kit/grpccodes"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/runtime/protoiface"
+	"net/http"
 )
 
-const (
-	resourceInfoDefaultOwner = "dapr-components"
-	errorInfoDefaultDomain   = "dapr.io"
-	errorInfoResonUnknown    = "UNKNOWN_REASON"
+var (
+	log = logger.NewLogger("dapr.kit")
 )
 
-var UnknownErrorReason = WithErrorReason(errorInfoResonUnknown, codes.Unknown)
+// Error implements the Error interface and the interface that complies with "google.golang.org/grpc/status".FromError().
+// It can be used to send errors to HTTP and gRPC servers, indicating the correct status code for each.
+type Error struct {
+	// Added error details. To see available details see:
+	// https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto
+	Details []proto.Message
 
-// ResourceInfo is meant to be used by Dapr components
-// to indicate the Type and Name.
-type ResourceInfo struct {
-	Type  string
-	Name  string
-	Owner string
-}
+	// Status code for gRPC responses.
+	GrpcCode grpcCodes.Code
 
-// Option allows passing additional information
-// to the Error struct.
-// See With* functions for further details.
-type Option func(*Error)
-
-// Error encapsulates error information
-// with additional details like:
-//   - http code
-//   - grpcStatus code
-//   - error reason
-//   - metadata information
-//   - optional resourceInfo (componenttype/name)
-type Error struct {
-	err            error
-	description    string
-	reason         string
-	httpCode       int
-	grpcStatusCode codes.Code
-	metadata       map[string]string
-	resourceInfo   *ResourceInfo
+	// Status code for HTTP responses.
+	HttpCode int
+
+	// Message is the human-readable error message.
+	Message string
+
+	// Tag is a string identifying the error, used with HTTP responses only.
+	Tag string
 }
 
 // New create a new Error using the supplied metadata and Options
-func New(err error, metadata map[string]string, options ...Option) *Error {
-	if err == nil {
-		return nil
-	}
-
+func New(grpcCode grpcCodes.Code, httpCode int, message string, tag string) *Error {
```

Comment:
Since `Error` implements the error interface, consider changing the return type to `error`:


```suggestion
func New(grpcCode grpcCodes.Code, httpCode int, message string, tag string) error {
```

Reply:
Im not sure I'm 100% following what you are suggesting. The point of standardizing the errors in this manner was to have a new error with all the status and details additions, if I switch to returning a standard lib error, then we lose all that additional error info. If you were suggesting making my `Error` struct `error` the IDE complains about the name, as that is the standard lib error.

Reply:
`error` is an interface, so you can return `*Error` that implements `error`.

In the caller, then, they can use `errors.As` to check if an `error` is of type `*Error`.

See: https://earthly.dev/blog/golang-errors/

Reply:
I think I see what you are saying. What are your thoughts on this: Here we leave the `New()` func to return the `*Error`, then inside Dapr, when calling this New func, It'll look like (note its returning the standard `error`:
```
func PubSubNotFound(name string, pubsubType string, metadata map[string]string) error {
	message := fmt.Sprintf("pubsub %s is not found", name)

	return kitErrors.New(
		grpcCodes.NotFound,
		http.StatusBadRequest,
		message,
		kitErrors.CodePrefixPubSub+kitErrors.CodeNotFound,
	).
		WithErrorInfo(kitErrors.CodePrefixPubSub+kitErrors.CodeNotFound, metadata).
		WithResourceInfo(pubsubType, name, "", message)
}
```

Where upon seeing this err inside dapr code, say `http/api.go` in my case, I can do as you mentioned ^ where I have a conditional checking the type like the following:

```
		var kitError *kitErrors.Error
		if errors.As(validationErr, &kitError) {
		...
		}
```
This way in dapr we see the `error` type, and can check the errType with the `errors.As` as you mentioned. 

Upon making these changes, I still see the extra err_details 👍🏻

Reply:
I am also +1 on returning the primitive `error` type on `New` if we can. This is to follow the Go language decision to treat all errors the "same" until a type check is relevant. This is to make the return `err` variable more portable given a shared base type. You can see this done in the stdlib where, for example, all `net/url` functions will return a `net/url.Error` as a primitive `error` even though it will probably be of the `Error` type.

Do we expect it to be the case that the kit `*Error` will want to be mutated after it has been initially constructed? I _think_ we don't.. which means it's equally important that the type is immutable by the caller. All fields should also then be made private, and values returned through exported methods (if they need to be exposed at all). To keep the "builder" style constructor which I think we want, we can create an intermediate `Builder` type that results in an immutable `error` type.

Reply:
In this case, it seems better to implement the Builder pattern and make Error immutable. It is a common design pattern when building an instance is "too complex".

Reply:
Could we have a situation where we need to create an error deeper in the code, and then extend it with more details (like resource info for example) somewhere on the api layer? I ran into a situation like this with checking for invalid keyname in state stores. The method that checks the keyname doesn't know the name of the state store we're working with, for example, which means we can't construct the ResourceInfo error detail in there, but we can extend the error once it gets to the api level. There are ways to go around this, of course, if we want to keep errors immutable, but it's probably worth considering.

Reply:
This PR has been updated accordingly: to include the builder pattern (see readme for example usage) and includes an `AddDetails` to the `Error` struct for adding details in runtime when needed.